Crash when moving an email to a folder since 5.734

**Describe the bug**
Since version 5.734 K9 sometimes crashes when I move an email to another folder (right after I click on the folder name). This doesn't happen every time though so it is hard to reproduce consistently, but the changelog for 5.734 noted many internal changes and asked to spot any new bugs. This seems to be related and I haven't seen a similar bug report.

**To Reproduce**
Steps to reproduce the behavior:
1. Select an email
2. Click on 'Move'
3. Select a folder
4. K9 crashes (I suddenly see my home screen without any error message or such)
5. Note: when I start K9 again, the email actually is moved to the new folder

**Expected behavior**
The email is moved and K9 doesn't crash

**Environment (please complete the following information):**
 - K-9 Mail version: 5.734
 - Android version: 11
 - Device: Samsung Galaxy S10
 - Account type: IMAP

**Logs**
Please take some time to [retrieve logs](https://github.com/k9mail/k-9/wiki/LoggingErrors) and attach them here:

@cketti right! I got a log of the crash. If you need main and system logs let me know. I still can't figure out exactly when/why this happens. I have a hunch that it is related to clicking on a link in an email and then clicking the back button to return to K9. If I then move the email to another folder it seems to happen?

```
--------- beginning of crash
06-17 03:11:01.253  5319  5319 E AndroidRuntime: FATAL EXCEPTION: main
06-17 03:11:01.253  5319  5319 E AndroidRuntime: Process: com.fsck.k9, PID: 5319
06-17 03:11:01.253  5319  5319 E AndroidRuntime: java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=1946015238, result=-1, data=Intent { (has extras) }} to activity {com.fsck.k9/com.fsck.k9.activity.MessageList}: java.lang.IllegalStateException: got an activity result that wasn't meant for us. this is a bug!
06-17 03:11:01.253  5319  5319 E AndroidRuntime: 	at android.app.ActivityThread.deliverResults(ActivityThread.java:5506)
06-17 03:11:01.253  5319  5319 E AndroidRuntime: 	at android.app.ActivityThread.handleSendResult(ActivityThread.java:5547)
06-17 03:11:01.253  5319  5319 E AndroidRuntime: 	at android.app.servertransaction.ActivityResultItem.execute(ActivityResultItem.java:51)
06-17 03:11:01.253  5319  5319 E AndroidRuntime: 	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
06-17 03:11:01.253  5319  5319 E AndroidRuntime: 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
06-17 03:11:01.253  5319  5319 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2307)
06-17 03:11:01.253  5319  5319 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:106)
06-17 03:11:01.253  5319  5319 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:246)
06-17 03:11:01.253  5319  5319 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:8512)
06-17 03:11:01.253  5319  5319 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
06-17 03:11:01.253  5319  5319 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
06-17 03:11:01.253  5319  5319 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
06-17 03:11:01.253  5319  5319 E AndroidRuntime: Caused by: java.lang.IllegalStateException: got an activity result that wasn't meant for us. this is a bug!
06-17 03:11:01.253  5319  5319 E AndroidRuntime: 	at com.fsck.k9.ui.messageview.MessageCryptoPresenter.onActivityResult(MessageCryptoPresenter.java:127)
06-17 03:11:01.253  5319  5319 E AndroidRuntime: 	at com.fsck.k9.ui.messageview.MessageViewFragment.onPendingIntentResult(MessageViewFragment.java:448)
06-17 03:11:01.253  5319  5319 E AndroidRuntime: 	at com.fsck.k9.activity.MessageList.onActivityResult(MessageList.kt:1444)
06-17 03:11:01.253  5319  5319 E AndroidRuntime: 	at android.app.Activity.dispatchActivityResult(Activity.java:8541)
06-17 03:11:01.253  5319  5319 E AndroidRuntime: 	at android.app.ActivityThread.deliverResults(ActivityThread.java:5499)
06-17 03:11:01.253  5319  5319 E AndroidRuntime: 	... 11 more
06-17 03:11:01.254  5319  5944 I MessagingController: moveOrCopyMessageSynchronous: source folder = 6, 1 messages, destination folder = 62, operation = MOVE
06-17 03:11:01.254  5319  5944 D MoveMessageOperations: Moving message [ID: 106214] to folder [ID: 62]
06-17 03:11:01.268  5319  5944 V LockableDatabase: LockableDatabase: Transaction ended, took 11 ms / com.fsck.k9.storage.messages.MoveMessageOperations.moveMessage(MoveMessageOperations.kt:20)
06-17 03:11:01.269  5319  5944 I MessagingController:  Command 'moveMessages' completed
06-17 03:11:01.269  5319  5944 I MessagingController: Running command 'processPendingCommands', seq = 2 (background priority)
06-17 03:11:01.270  5319  5944 D MessagingController: Processing pending command 'move_or_copy'
06-17 03:11:01.274  5319  5319 I Process : Sending signal. PID: 5319 SIG: 9
```

@cketti I am able to reproduce the log seen above, with the 5.734 version. As well as one other failure mode.
Logs attached:
```
$ cat k9-log.txt
--------- beginning of main
06-20 21:03:16.394 30592 30592 I InputTransport: Create ARC handle: 0x7660cb3380
06-20 21:03:16.395 30592 30592 V PhoneWindow: DecorView setVisiblity: visibility = 0, Parent = android.view.ViewRootImpl@4335cfd, this = DecorView@8e1ddf2[MessageList]
06-20 21:03:16.427 30592 30592 E GraphicExt: Can't load libboost_ext_fwk
06-20 21:03:16.427 30592 30592 E GraphicExt: GraphicExtModuleLoader::CreateGraphicExtInstance false
06-20 21:03:16.428 30592 30616 I GPUD    : @gpudInitialize: successfully initialized with GL, dbg=0 mmdump_dbg=0 mmpath_dbg=0
06-20 21:03:16.444 30592 30616 D Surface : Surface::connect(this=0x774df1c000,api=1)
06-20 21:03:16.445 30592 30616 D mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
06-20 21:03:16.445 30592 30616 D Surface : Surface::setBufferCount(this=0x774df1c000,bufferCount=3)
06-20 21:03:16.445 30592 30616 D Surface : Surface::allocateBuffers(this=0x774df1c000)
06-20 21:03:16.449 30592 30616 W Gralloc3: mapper 3.x is not supported
06-20 21:03:16.452 30592 30616 E gralloc : Arm Module v1.0
06-20 21:03:16.454 30592 30616 E ion     : ioctl c0044901 failed with code -1: Invalid argument
06-20 21:03:17.436 30592 30592 I ViewGroup: w:1432 h:1436
06-20 21:03:17.439 30592 30592 I chatty  : uid=10003(com.fsck.k9) identical 10 lines
06-20 21:03:17.439 30592 30592 I ViewGroup: w:1432 h:1436
06-20 21:03:17.912 30592 30592 I taoqiong_back_press: onPause getComponentName().getClassName()：com.fsck.k9.activity.MessageList
06-20 21:03:17.922 30592 30592 W ActivityThread: handleWindowVisibility: no activity for token android.os.BinderProxy@d75e935
06-20 21:03:17.943 30592 30592 I taoqiong_back_press: onResume getComponentName().getClassName()：com.fsck.k9.ui.choosefolder.ChooseFolderActivity
06-20 21:03:17.944 30592 30592 V PhoneWindow: DecorView setVisiblity: visibility = 4, Parent = null, this = DecorView@9e1fd1e[]
06-20 21:03:17.946 30592 30592 D ViewRootImpl[ChooseFolderActivity]: hardware acceleration = true , fakeHwAccelerated = false, sRendererDisabled = false, forceHwAccelerated = false, sSystemRendererDisabled = false
06-20 21:03:17.950 30592 30592 I InputTransport: Create ARC handle: 0x7628b6dca0
06-20 21:03:17.951 30592 30592 V PhoneWindow: DecorView setVisiblity: visibility = 0, Parent = android.view.ViewRootImpl@1542615, this = DecorView@9e1fd1e[ChooseFolderActivity]
--------- beginning of system
06-20 21:03:17.953 30592 30592 W ViewRootImpl[MessageList]: Cancelling event due to no window focus: KeyEvent { action=ACTION_UP, keyCode=KEYCODE_Y, scanCode=21, metaState=0, flags=0x28, repeatCount=0, eventTime=46047473, downTime=46047424, deviceId=0, source=0x101, displayId=0 }
06-20 21:03:17.953 30592 30592 I chatty  : uid=10003(com.fsck.k9) identical 1 line
06-20 21:03:17.953 30592 30592 W ViewRootImpl[MessageList]: Cancelling event due to no window focus: KeyEvent { action=ACTION_UP, keyCode=KEYCODE_Y, scanCode=21, metaState=0, flags=0x28, repeatCount=0, eventTime=46047473, downTime=46047424, deviceId=0, source=0x101, displayId=0 }
06-20 21:03:17.962 30592 30592 E GraphicExt: GraphicExtModuleLoader::CreateGraphicExtInstance false
06-20 21:03:17.962 30592 30616 D Surface : Surface::connect(this=0x7628a32000,api=1)
06-20 21:03:17.963 30592 30616 D mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
06-20 21:03:17.963 30592 30616 D Surface : Surface::setBufferCount(this=0x7628a32000,bufferCount=3)
06-20 21:03:17.963 30592 30616 D Surface : Surface::allocateBuffers(this=0x7628a32000)
06-20 21:03:17.974 30592 30592 W ViewRootImpl[MessageList]: Cancelling event due to no window focus: KeyEvent { action=ACTION_UP, keyCode=KEYCODE_Y, scanCode=21, metaState=0, flags=0x28, repeatCount=0, eventTime=46047473, downTime=46047424, deviceId=0, source=0x101, displayId=0 }
06-20 21:03:17.974 30592 30592 I chatty  : uid=10003(com.fsck.k9) identical 4 lines
06-20 21:03:17.974 30592 30592 W ViewRootImpl[MessageList]: Cancelling event due to no window focus: KeyEvent { action=ACTION_UP, keyCode=KEYCODE_Y, scanCode=21, metaState=0, flags=0x28, repeatCount=0, eventTime=46047473, downTime=46047424, deviceId=0, source=0x101, displayId=0 }
06-20 21:03:18.401 30592 30616 D Surface : Surface::disconnect(this=0x774df1c000,api=1)
06-20 21:03:18.412 30592 30592 V PhoneWindow: DecorView setVisiblity: visibility = 4, Parent = android.view.ViewRootImpl@4335cfd, this = DecorView@8e1ddf2[MessageList]
06-20 21:03:19.468 30592 30592 I ViewGroup: w:1432 h:1436
06-20 21:03:19.469 30592 30592 I chatty  : uid=10003(com.fsck.k9) identical 6 lines
06-20 21:03:19.469 30592 30592 I ViewGroup: w:1432 h:1436
06-20 21:03:20.166 30592 30592 I taoqiong_back_press: getComponentName().getClassName()：com.fsck.k9.ui.choosefolder.ChooseFolderActivity
06-20 21:03:20.170 30592 30592 I taoqiong_back_press: onPause getComponentName().getClassName()：com.fsck.k9.ui.choosefolder.ChooseFolderActivity
06-20 21:03:20.180 30592 30592 I taoqiong_back_press: onResume getComponentName().getClassName()：com.fsck.k9.activity.MessageList
06-20 21:03:20.180 30592 30592 V PhoneWindow: DecorView setVisiblity: visibility = 0, Parent = android.view.ViewRootImpl@4335cfd, this = DecorView@8e1ddf2[MessageList]
06-20 21:03:20.187 30592 30592 E GraphicExt: GraphicExtModuleLoader::CreateGraphicExtInstance false
06-20 21:03:20.188 30592 30616 D Surface : Surface::connect(this=0x7628de8000,api=1)
06-20 21:03:20.189 30592 30616 D mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
06-20 21:03:20.189 30592 30616 D Surface : Surface::setBufferCount(this=0x7628de8000,bufferCount=3)
06-20 21:03:20.189 30592 30616 D Surface : Surface::allocateBuffers(this=0x7628de8000)
06-20 21:03:20.622 30592 30616 D Surface : Surface::disconnect(this=0x7628a32000,api=1)
06-20 21:03:20.629 30592 30592 V PhoneWindow: DecorView setVisiblity: visibility = 4, Parent = android.view.ViewRootImpl@1542615, this = DecorView@9e1fd1e[ChooseFolderActivity]
06-20 21:03:20.630 30592 30592 D View    : [Warning] assignParent to null: this = DecorView@9e1fd1e[ChooseFolderActivity]
06-20 21:03:20.630 30592 30592 I InputTransport: Destroy ARC handle: 0x7628b6dca0
06-20 21:03:21.085 30592 30592 I taoqiong_back_press: getComponentName().getClassName()：com.fsck.k9.activity.MessageList
06-20 21:03:21.092 30592 30592 I taoqiong_back_press: onPause getComponentName().getClassName()：com.fsck.k9.activity.MessageList
06-20 21:03:21.782 30592 30616 D Surface : Surface::disconnect(this=0x7628de8000,api=1)
06-20 21:03:21.805 30592 30616 I GED     : ged_boost_gpu_freq, level 100, eOrigin 2, final_idx 15, oppidx_max 15, oppidx_min 0
06-20 21:03:21.811 30592 30592 V PhoneWindow: DecorView setVisiblity: visibility = 4, Parent = android.view.ViewRootImpl@4335cfd, this = DecorView@8e1ddf2[MessageList]
06-20 21:03:21.834 30592 30592 D View    : [Warning] assignParent to null: this = DecorView@8e1ddf2[MessageList]
06-20 21:03:21.835 30592 30592 I InputTransport: Destroy ARC handle: 0x7660cb3380
06-20 21:03:22.157 30592 30592 W ActivityThread: handleWindowVisibility: no activity for token android.os.BinderProxy@c4ea86f
06-20 21:03:22.255 30592 30592 I taoqiong_back_press: onResume getComponentName().getClassName()：com.fsck.k9.activity.MessageList
06-20 21:03:22.256 30592 30592 V PhoneWindow: DecorView setVisiblity: visibility = 4, Parent = null, this = DecorView@6eaf172[]
06-20 21:03:22.258 30592 30592 D ViewRootImpl[MessageList]: hardware acceleration = true , fakeHwAccelerated = false, sRendererDisabled = false, forceHwAccelerated = false, sSystemRendererDisabled = false
06-20 21:03:22.263 30592 30592 I InputTransport: Create ARC handle: 0x7628deac60
06-20 21:03:22.264 30592 30592 V PhoneWindow: DecorView setVisiblity: visibility = 0, Parent = android.view.ViewRootImpl@91a9579, this = DecorView@6eaf172[MessageList]
06-20 21:03:22.282 30592 30592 E GraphicExt: GraphicExtModuleLoader::CreateGraphicExtInstance false
06-20 21:03:22.283 30592 30616 D Surface : Surface::connect(this=0x7660c81000,api=1)
06-20 21:03:22.284 30592 30616 D mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
06-20 21:03:22.284 30592 30616 D Surface : Surface::setBufferCount(this=0x7660c81000,bufferCount=3)
06-20 21:03:22.284 30592 30616 D Surface : Surface::allocateBuffers(this=0x7660c81000)
06-20 21:03:24.240 30592 30592 I WebViewFactory: Loading com.google.android.webview version 91.0.4472.101 (code 447210133)
06-20 21:03:24.341 30592 30592 I cr_LibraryLoader: Loaded native library version number "91.0.4472.101"
06-20 21:03:24.342 30592 30592 I cr_CachingUmaRecorder: Flushed 6 samples from 6 histograms.
06-20 21:03:24.471 30592 30719 W cr_media: Requires BLUETOOTH permission
06-20 21:03:24.493 30592 30731 I GED     : ged_boost_gpu_freq, level 100, eOrigin 2, final_idx 15, oppidx_max 15, oppidx_min 0
06-20 21:03:24.849 30592 30592 I ViewGroup: w:1432 h:1436
06-20 21:03:24.855 30592 30592 I chatty  : uid=10003(com.fsck.k9) identical 11 lines
06-20 21:03:24.855 30592 30592 I ViewGroup: w:1432 h:1436
06-20 21:03:25.225 30592 30592 I taoqiong_back_press: onPause getComponentName().getClassName()：com.fsck.k9.activity.MessageList
06-20 21:03:25.241 30592 30592 W ActivityThread: handleWindowVisibility: no activity for token android.os.BinderProxy@80035e
06-20 21:03:25.261 30592 30592 I taoqiong_back_press: onResume getComponentName().getClassName()：com.fsck.k9.ui.choosefolder.ChooseFolderActivity
06-20 21:03:25.261 30592 30592 V PhoneWindow: DecorView setVisiblity: visibility = 4, Parent = null, this = DecorView@21c4d1f[]
06-20 21:03:25.262 30592 30592 D ViewRootImpl[ChooseFolderActivity]: hardware acceleration = true , fakeHwAccelerated = false, sRendererDisabled = false, forceHwAccelerated = false, sSystemRendererDisabled = false
06-20 21:03:25.265 30592 30592 I InputTransport: Create ARC handle: 0x765920f700
06-20 21:03:25.265 30592 30592 V PhoneWindow: DecorView setVisiblity: visibility = 0, Parent = android.view.ViewRootImpl@5cc72ca, this = DecorView@21c4d1f[ChooseFolderActivity]
06-20 21:03:25.268 30592 30592 W ViewRootImpl[MessageList]: Cancelling event due to no window focus: KeyEvent { action=ACTION_UP, keyCode=KEYCODE_Y, scanCode=21, metaState=0, flags=0x28, repeatCount=0, eventTime=46054789, downTime=46054728, deviceId=0, source=0x101, displayId=0 }
06-20 21:03:25.268 30592 30592 I chatty  : uid=10003(com.fsck.k9) identical 7 lines
06-20 21:03:25.268 30592 30592 W ViewRootImpl[MessageList]: Cancelling event due to no window focus: KeyEvent { action=ACTION_UP, keyCode=KEYCODE_Y, scanCode=21, metaState=0, flags=0x28, repeatCount=0, eventTime=46054789, downTime=46054728, deviceId=0, source=0x101, displayId=0 }
06-20 21:03:25.275 30592 30592 E GraphicExt: GraphicExtModuleLoader::CreateGraphicExtInstance false
06-20 21:03:25.275 30592 30616 D Surface : Surface::connect(this=0x76232ce000,api=1)
06-20 21:03:25.276 30592 30616 D mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
06-20 21:03:25.276 30592 30616 D Surface : Surface::setBufferCount(this=0x76232ce000,bufferCount=3)
06-20 21:03:25.276 30592 30616 D Surface : Surface::allocateBuffers(this=0x76232ce000)
06-20 21:03:25.719 30592 30616 D Surface : Surface::disconnect(this=0x7660c81000,api=1)
06-20 21:03:25.733 30592 30592 V PhoneWindow: DecorView setVisiblity: visibility = 4, Parent = android.view.ViewRootImpl@91a9579, this = DecorView@6eaf172[MessageList]
06-20 21:03:25.912 30592 30592 I ViewGroup: w:1432 h:1436
06-20 21:03:25.913 30592 30592 I chatty  : uid=10003(com.fsck.k9) identical 6 lines
06-20 21:03:25.914 30592 30592 I ViewGroup: w:1432 h:1436
06-20 21:03:26.280 30592 30592 I taoqiong_back_press: getComponentName().getClassName()：com.fsck.k9.ui.choosefolder.ChooseFolderActivity
06-20 21:03:26.286 30592 30592 I taoqiong_back_press: onPause getComponentName().getClassName()：com.fsck.k9.ui.choosefolder.ChooseFolderActivity
06-20 21:03:26.293 30592 30592 D AndroidRuntime: Shutting down VM
--------- beginning of crash
06-20 21:03:26.294 30592 30592 E AndroidRuntime: FATAL EXCEPTION: main
06-20 21:03:26.294 30592 30592 E AndroidRuntime: Process: com.fsck.k9, PID: 30592
06-20 21:03:26.294 30592 30592 E AndroidRuntime: java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=454665778, result=0, data=null} to activity {com.fsck.k9/com.fsck.k9.activity.MessageList}: java.lang.IllegalStateException: got an activity result that wasn't meant for us. this is a bug!
06-20 21:03:26.294 30592 30592 E AndroidRuntime: 	at android.app.ActivityThread.deliverResults(ActivityThread.java:4873)
06-20 21:03:26.294 30592 30592 E AndroidRuntime: 	at android.app.ActivityThread.handleSendResult(ActivityThread.java:4914)
06-20 21:03:26.294 30592 30592 E AndroidRuntime: 	at android.app.servertransaction.ActivityResultItem.execute(ActivityResultItem.java:51)
06-20 21:03:26.294 30592 30592 E AndroidRuntime: 	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
06-20 21:03:26.294 30592 30592 E AndroidRuntime: 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
06-20 21:03:26.294 30592 30592 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2041)
06-20 21:03:26.294 30592 30592 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:107)
06-20 21:03:26.294 30592 30592 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:214)
06-20 21:03:26.294 30592 30592 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7386)
06-20 21:03:26.294 30592 30592 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
06-20 21:03:26.294 30592 30592 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
06-20 21:03:26.294 30592 30592 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:980)
06-20 21:03:26.294 30592 30592 E AndroidRuntime: Caused by: java.lang.IllegalStateException: got an activity result that wasn't meant for us. this is a bug!
06-20 21:03:26.294 30592 30592 E AndroidRuntime: 	at com.fsck.k9.ui.messageview.MessageCryptoPresenter.onActivityResult(MessageCryptoPresenter.java:127)
06-20 21:03:26.294 30592 30592 E AndroidRuntime: 	at com.fsck.k9.ui.messageview.MessageViewFragment.onPendingIntentResult(MessageViewFragment.java:448)
06-20 21:03:26.294 30592 30592 E AndroidRuntime: 	at com.fsck.k9.activity.MessageList.onActivityResult(MessageList.kt:1444)
06-20 21:03:26.294 30592 30592 E AndroidRuntime: 	at android.app.Activity.dispatchActivityResult(Activity.java:8164)
06-20 21:03:26.294 30592 30592 E AndroidRuntime: 	at android.app.ActivityThread.deliverResults(ActivityThread.java:4866)
06-20 21:03:26.294 30592 30592 E AndroidRuntime: 	... 11 more
```
And then this one:
```
$ cat k9-log2.txt
--------- beginning of main
06-20 21:07:04.063  1670  1670 E com.fsck.k9: Not starting debugger since process cannot load the jdwp agent.
06-20 21:07:04.147  1670  1670 D NetworkSecurityConfig: No Network Security Config specified, using platform default
06-20 21:07:04.291  1670  1670 I taoqiong_back_press: onResume getComponentName().getClassName()：com.fsck.k9.activity.MessageList
06-20 21:07:04.294  1670  1670 I SurfaceFactory: [static] sSurfaceFactory = com.mediatek.view.impl.SurfaceFactoryImpl@7db25c1
06-20 21:07:04.299  1670  1670 D ViewRootImpl[MessageList]: hardware acceleration = true , fakeHwAccelerated = false, sRendererDisabled = false, forceHwAccelerated = false, sSystemRendererDisabled = false
06-20 21:07:04.304  1670  1670 I InputTransport: Create ARC handle: 0x7660aeb5c0
06-20 21:07:04.305  1670  1670 V PhoneWindow: DecorView setVisiblity: visibility = 0, Parent = android.view.ViewRootImpl@4335cfd, this = DecorView@8e1ddf2[MessageList]
06-20 21:07:04.331  1670  1670 E GraphicExt: Can't load libboost_ext_fwk
06-20 21:07:04.331  1670  1670 E GraphicExt: GraphicExtModuleLoader::CreateGraphicExtInstance false
06-20 21:07:04.332  1670  1694 I GPUD    : @gpudInitialize: successfully initialized with GL, dbg=0 mmdump_dbg=0 mmpath_dbg=0
06-20 21:07:04.344  1670  1694 D Surface : Surface::connect(this=0x774df1d000,api=1)
06-20 21:07:04.345  1670  1694 D mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
06-20 21:07:04.345  1670  1694 D Surface : Surface::setBufferCount(this=0x774df1d000,bufferCount=3)
06-20 21:07:04.345  1670  1694 D Surface : Surface::allocateBuffers(this=0x774df1d000)
06-20 21:07:04.349  1670  1694 W Gralloc3: mapper 3.x is not supported
06-20 21:07:04.350  1670  1694 E gralloc : Arm Module v1.0
06-20 21:07:04.351  1670  1694 E ion     : ioctl c0044901 failed with code -1: Invalid argument
06-20 21:07:06.240  1670  1670 I WebViewFactory: Loading com.google.android.webview version 91.0.4472.101 (code 447210133)
06-20 21:07:06.302  1670  1670 I cr_LibraryLoader: Loaded native library version number "91.0.4472.101"
06-20 21:07:06.303  1670  1670 I cr_CachingUmaRecorder: Flushed 6 samples from 6 histograms.
06-20 21:07:06.410  1670  1763 W cr_media: Requires BLUETOOTH permission
06-20 21:07:06.428  1670  1775 I GED     : ged_boost_gpu_freq, level 100, eOrigin 2, final_idx 15, oppidx_max 15, oppidx_min 0
06-20 21:07:06.958  1670  1670 I ViewGroup: w:1432 h:1436
06-20 21:07:06.962  1670  1670 I chatty  : uid=10003(com.fsck.k9) identical 11 lines
06-20 21:07:06.962  1670  1670 I ViewGroup: w:1432 h:1436
06-20 21:07:07.110  1670  1670 I taoqiong_back_press: onPause getComponentName().getClassName()：com.fsck.k9.activity.MessageList
06-20 21:07:07.125  1670  1670 W ActivityThread: handleWindowVisibility: no activity for token android.os.BinderProxy@112d65b
06-20 21:07:07.144  1670  1670 I taoqiong_back_press: onResume getComponentName().getClassName()：com.fsck.k9.ui.choosefolder.ChooseFolderActivity
06-20 21:07:07.144  1670  1670 V PhoneWindow: DecorView setVisiblity: visibility = 4, Parent = null, this = DecorView@2347ab1[]
06-20 21:07:07.145  1670  1670 D ViewRootImpl[ChooseFolderActivity]: hardware acceleration = true , fakeHwAccelerated = false, sRendererDisabled = false, forceHwAccelerated = false, sSystemRendererDisabled = false
06-20 21:07:07.148  1670  1670 I InputTransport: Create ARC handle: 0x7628c33ae0
06-20 21:07:07.149  1670  1670 V PhoneWindow: DecorView setVisiblity: visibility = 0, Parent = android.view.ViewRootImpl@8a5e304, this = DecorView@2347ab1[ChooseFolderActivity]
--------- beginning of system
06-20 21:07:07.151  1670  1670 W ViewRootImpl[MessageList]: Cancelling event due to no window focus: KeyEvent { action=ACTION_UP, keyCode=KEYCODE_Y, scanCode=21, metaState=0, flags=0x28, repeatCount=0, eventTime=46276674, downTime=46276628, deviceId=0, source=0x101, displayId=0 }
06-20 21:07:07.152  1670  1670 I chatty  : uid=10003(com.fsck.k9) identical 7 lines
06-20 21:07:07.152  1670  1670 W ViewRootImpl[MessageList]: Cancelling event due to no window focus: KeyEvent { action=ACTION_UP, keyCode=KEYCODE_Y, scanCode=21, metaState=0, flags=0x28, repeatCount=0, eventTime=46276674, downTime=46276628, deviceId=0, source=0x101, displayId=0 }
06-20 21:07:07.159  1670  1670 E GraphicExt: GraphicExtModuleLoader::CreateGraphicExtInstance false
06-20 21:07:07.159  1670  1694 D Surface : Surface::connect(this=0x7628c4d000,api=1)
06-20 21:07:07.160  1670  1694 D mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
06-20 21:07:07.160  1670  1694 D Surface : Surface::setBufferCount(this=0x7628c4d000,bufferCount=3)
06-20 21:07:07.161  1670  1694 D Surface : Surface::allocateBuffers(this=0x7628c4d000)
06-20 21:07:07.602  1670  1694 D Surface : Surface::disconnect(this=0x774df1d000,api=1)
06-20 21:07:07.614  1670  1670 V PhoneWindow: DecorView setVisiblity: visibility = 4, Parent = android.view.ViewRootImpl@4335cfd, this = DecorView@8e1ddf2[MessageList]
06-20 21:07:07.797  1670  1670 I ViewGroup: w:1432 h:1436
06-20 21:07:07.798  1670  1670 I chatty  : uid=10003(com.fsck.k9) identical 6 lines
06-20 21:07:07.798  1670  1670 I ViewGroup: w:1432 h:1436
06-20 21:07:08.067  1670  1670 I taoqiong_back_press: getComponentName().getClassName()：com.fsck.k9.ui.choosefolder.ChooseFolderActivity
06-20 21:07:08.071  1670  1670 I taoqiong_back_press: onPause getComponentName().getClassName()：com.fsck.k9.ui.choosefolder.ChooseFolderActivity
06-20 21:07:08.078  1670  1670 D AndroidRuntime: Shutting down VM
--------- beginning of crash
06-20 21:07:08.078  1670  1670 E AndroidRuntime: FATAL EXCEPTION: main
06-20 21:07:08.078  1670  1670 E AndroidRuntime: Process: com.fsck.k9, PID: 1670
06-20 21:07:08.078  1670  1670 E AndroidRuntime: java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=1548070286, result=0, data=null} to activity {com.fsck.k9/com.fsck.k9.activity.MessageList}: java.lang.NullPointerException: Attempt to invoke virtual method 'void com.fsck.k9.ui.crypto.MessageCryptoHelper.onActivityResult(int, int, android.content.Intent)' on a null object reference
06-20 21:07:08.078  1670  1670 E AndroidRuntime: 	at android.app.ActivityThread.deliverResults(ActivityThread.java:4873)
06-20 21:07:08.078  1670  1670 E AndroidRuntime: 	at android.app.ActivityThread.handleSendResult(ActivityThread.java:4914)
06-20 21:07:08.078  1670  1670 E AndroidRuntime: 	at android.app.servertransaction.ActivityResultItem.execute(ActivityResultItem.java:51)
06-20 21:07:08.078  1670  1670 E AndroidRuntime: 	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
06-20 21:07:08.078  1670  1670 E AndroidRuntime: 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
06-20 21:07:08.078  1670  1670 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2041)
06-20 21:07:08.078  1670  1670 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:107)
06-20 21:07:08.078  1670  1670 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:214)
06-20 21:07:08.078  1670  1670 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7386)
06-20 21:07:08.078  1670  1670 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
06-20 21:07:08.078  1670  1670 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
06-20 21:07:08.078  1670  1670 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:980)
06-20 21:07:08.078  1670  1670 E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void com.fsck.k9.ui.crypto.MessageCryptoHelper.onActivityResult(int, int, android.content.Intent)' on a null object reference
06-20 21:07:08.078  1670  1670 E AndroidRuntime: 	at com.fsck.k9.activity.MessageLoaderHelper.onActivityResult(MessageLoaderHelper.java:194)
06-20 21:07:08.078  1670  1670 E AndroidRuntime: 	at com.fsck.k9.ui.messageview.MessageViewFragment.onPendingIntentResult(MessageViewFragment.java:442)
06-20 21:07:08.078  1670  1670 E AndroidRuntime: 	at com.fsck.k9.activity.MessageList.onActivityResult(MessageList.kt:1444)
06-20 21:07:08.078  1670  1670 E AndroidRuntime: 	at android.app.Activity.dispatchActivityResult(Activity.java:8164)
06-20 21:07:08.078  1670  1670 E AndroidRuntime: 	at android.app.ActivityThread.deliverResults(ActivityThread.java:4866)
06-20 21:07:08.078  1670  1670 E AndroidRuntime: 	... 11 more
```
In my reprodcution, I am using a Unihertz Titan phone, which has a physical keyboard.
I launch k9. Open the first email message form the unified inbox and press the 'Y' key to copy the message to a folder.
Then press the 'back' key, to cancel that action.
When it doesn't reproduce the failure, I exit the app and try again.
After some number of attempts I am able to get the first failure.  And after that second one too.
I did some experiments, and I found that the following change stops the failure.
```
diff --git a/app/ui/legacy/src/main/java/com/fsck/k9/activity/MessageLoaderHelper.java b/app/ui/legacy/src/main/java/com/fsck/k9/activity/MessageLoaderHelper.java
index 229b48103..1233ead09 100644
--- a/app/ui/legacy/src/main/java/com/fsck/k9/activity/MessageLoaderHelper.java
+++ b/app/ui/legacy/src/main/java/com/fsck/k9/activity/MessageLoaderHelper.java
@@ -191,7 +191,11 @@ public class MessageLoaderHelper {
 
     @UiThread
     public void onActivityResult(int requestCode, int resultCode, Intent data) {
-        messageCryptoHelper.onActivityResult(requestCode, resultCode, data);
+        if (messageCryptoHelper != null) {
+            messageCryptoHelper.onActivityResult(requestCode, resultCode, data);
+        } else {
+            Timber.d("Ignoring ActivityResult, messageCryptoHelper is null");
+        }
     }
 
 
diff --git a/app/ui/legacy/src/main/java/com/fsck/k9/ui/messageview/MessageCryptoPresenter.java b/app/ui/legacy/src/main/java/com/fsck/k9/ui/messageview/MessageCryptoPresenter.java
index 6a192e5ea..250c9aa20 100644
--- a/app/ui/legacy/src/main/java/com/fsck/k9/ui/messageview/MessageCryptoPresenter.java
+++ b/app/ui/legacy/src/main/java/com/fsck/k9/ui/messageview/MessageCryptoPresenter.java
@@ -124,6 +124,10 @@ public class MessageCryptoPresenter implements OnCryptoClickListener {
         } else if (requestCode == REQUEST_CODE_SECURITY_WARNING) {
             messageCryptoMvpView.redisplayMessage();
         } else {
+            if (resultCode == Activity.RESULT_CANCELED) {
+                Timber.d("ignoring canceled activity that wasn't meant for us.");
+                return;
+            }
             throw new IllegalStateException("got an activity result that wasn't meant for us. this is a bug!");
         }
     }
```
I am not familiar with the code, so I don't know if this is a good change, or if it is is only covering up the root cause of failure.

--- a/app/ui/legacy/src/main/java/com/fsck/k9/activity/MessageCompose.java
+++ b/app/ui/legacy/src/main/java/com/fsck/k9/activity/MessageCompose.java
@@ -163,6 +163,7 @@ public class MessageCompose extends K9Activity implements OnClickListener,
     public static final int MSG_SAVED_DRAFT = 4;
     private static final int MSG_DISCARDED_DRAFT = 5;
 
+    private static final int REQUEST_CODE_MASK = 0xFFFF0000;
     private static final int REQUEST_MASK_RECIPIENT_PRESENTER = (1 << 8);
     private static final int REQUEST_MASK_LOADER_HELPER = (1 << 9);
     private static final int REQUEST_MASK_ATTACHMENT_PRESENTER = (1 << 10);
@@ -831,33 +832,36 @@ public class MessageCompose extends K9Activity implements OnClickListener,
     protected void onActivityResult(int requestCode, int resultCode, Intent data) {
         isInSubActivity = false;
 
-        if ((requestCode & REQUEST_MASK_MESSAGE_BUILDER) == REQUEST_MASK_MESSAGE_BUILDER) {
-            requestCode ^= REQUEST_MASK_MESSAGE_BUILDER;
-            if (currentMessageBuilder == null) {
-                Timber.e("Got a message builder activity result for no message builder, " +
-                        "this is an illegal state!");
+        // Only if none of the high 16 bits are set it might be one of our request codes
+        if ((requestCode & REQUEST_CODE_MASK) == 0) {
+            if ((requestCode & REQUEST_MASK_MESSAGE_BUILDER) == REQUEST_MASK_MESSAGE_BUILDER) {
+                requestCode ^= REQUEST_MASK_MESSAGE_BUILDER;
+                if (currentMessageBuilder == null) {
+                    Timber.e("Got a message builder activity result for no message builder, " +
+                            "this is an illegal state!");
+                    return;
+                }
+                currentMessageBuilder.onActivityResult(requestCode, resultCode, data, this);
                 return;
             }
-            currentMessageBuilder.onActivityResult(requestCode, resultCode, data, this);
-            return;
-        }
 
-        if ((requestCode & REQUEST_MASK_RECIPIENT_PRESENTER) == REQUEST_MASK_RECIPIENT_PRESENTER) {
-            requestCode ^= REQUEST_MASK_RECIPIENT_PRESENTER;
-            recipientPresenter.onActivityResult(requestCode, resultCode, data);
-            return;
-        }
+            if ((requestCode & REQUEST_MASK_RECIPIENT_PRESENTER) == REQUEST_MASK_RECIPIENT_PRESENTER) {
+                requestCode ^= REQUEST_MASK_RECIPIENT_PRESENTER;
+                recipientPresenter.onActivityResult(requestCode, resultCode, data);
+                return;
+            }
 
-        if ((requestCode & REQUEST_MASK_LOADER_HELPER) == REQUEST_MASK_LOADER_HELPER) {
-            requestCode ^= REQUEST_MASK_LOADER_HELPER;
-            messageLoaderHelper.onActivityResult(requestCode, resultCode, data);
-            return;
-        }
+            if ((requestCode & REQUEST_MASK_LOADER_HELPER) == REQUEST_MASK_LOADER_HELPER) {
+                requestCode ^= REQUEST_MASK_LOADER_HELPER;
+                messageLoaderHelper.onActivityResult(requestCode, resultCode, data);
+                return;
+            }
 
-        if ((requestCode & REQUEST_MASK_ATTACHMENT_PRESENTER) == REQUEST_MASK_ATTACHMENT_PRESENTER) {
-            requestCode ^= REQUEST_MASK_ATTACHMENT_PRESENTER;
-            attachmentPresenter.onActivityResult(resultCode, requestCode, data);
-            return;
+            if ((requestCode & REQUEST_MASK_ATTACHMENT_PRESENTER) == REQUEST_MASK_ATTACHMENT_PRESENTER) {
+                requestCode ^= REQUEST_MASK_ATTACHMENT_PRESENTER;
+                attachmentPresenter.onActivityResult(resultCode, requestCode, data);
+                return;
+            }
         }
 
         super.onActivityResult(requestCode, resultCode, data);

--- a/app/ui/legacy/src/main/java/com/fsck/k9/activity/MessageList.kt
+++ b/app/ui/legacy/src/main/java/com/fsck/k9/activity/MessageList.kt
@@ -1421,15 +1421,24 @@ open class MessageList :
         flagsValues: Int,
         extraFlags: Int
     ) {
-        val modifiedRequestCode = requestCode or REQUEST_MASK_PENDING_INTENT
+        // If any of the high 16 bits are set it is not one of our request codes
+        if (requestCode and REQUEST_CODE_MASK != 0) {
+            super.startIntentSenderForResult(intent, requestCode, fillInIntent, flagsMask, flagsValues, extraFlags)
+            return
+        }
+
+        val modifiedRequestCode = requestCode or REQUEST_FLAG_PENDING_INTENT
         super.startIntentSenderForResult(intent, modifiedRequestCode, fillInIntent, flagsMask, flagsValues, extraFlags)
     }
 
     override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
         super.onActivityResult(requestCode, resultCode, data)
 
-        if (requestCode and REQUEST_MASK_PENDING_INTENT == REQUEST_MASK_PENDING_INTENT) {
-            val originalRequestCode = requestCode xor REQUEST_MASK_PENDING_INTENT
+        // If any of the high 16 bits are set it is not one of our request codes
+        if (requestCode and REQUEST_CODE_MASK != 0) return
+
+        if (requestCode and REQUEST_FLAG_PENDING_INTENT != 0) {
+            val originalRequestCode = requestCode xor REQUEST_FLAG_PENDING_INTENT
             if (messageViewFragment != null) {
                 messageViewFragment!!.onPendingIntentResult(originalRequestCode, resultCode, data)
             }
@@ -1554,7 +1563,8 @@ open class MessageList :
         private const val PREVIOUS = 1
         private const val NEXT = 2
 
-        const val REQUEST_MASK_PENDING_INTENT = 1 shl 15
+        private const val REQUEST_CODE_MASK = 0xFFFF0000.toInt()
+        private const val REQUEST_FLAG_PENDING_INTENT = 1 shl 15
 
         private val defaultFolderProvider: DefaultFolderProvider by inject()
 

